LiS/man/man2/getpmsg.2s.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) $Id: getpmsg.2s.man,v 1.1.4.4 2008-04-29 08:33:17 brian Exp $
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified $Date: 2008-04-29 08:33:17 $ by $Author: brian $
.\"
.\" -------------------------------------------------------------------------
.\"
.\" $Log: getpmsg.2s.man,v $
.\" Revision 1.1.4.4  2008-04-29 08:33:17  brian
.\" - update headers for Affero release
.\"
.\" Revision 1.1.4.3  2006/04/05 09:42:18  brian
.\" - moved .so line and updated headers
.\"
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database lis.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so lis.macros
.\"
.\"
.TH GETPMSG 2 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ System Calls"
.SH NAME
getpmsg \- get next message off a stream
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <stropts.h>
.HP 8
.BI "int " retval " = getpmsg(int " fd ", struct strbuf *" ctlptr ", struct strbuf *" dataptr ", int *" bandp ", int *" flagsp );
.\"
.\"
.SH ARGUMENTS
.PP
.TP
.B fd
the
.I STREAMS
file descriptor from which to get a message.
.TP
.B ctlptr
a pointer to a
.B strbuf
structure describing the buffer to which to receive any control part of the message.
.TP
.B dataptr
a pointer to a
.B strbuf
structure describing the buffer to which to receive any data part of the message.
.TP
.B bandp
a pointer to an integer into which to receive the band number associated with the message.
.TP
.B flagsp
a pointer to a flags integer into which to receive any flags associated with the message.
.\"
.\"
.SH DESCRIPTION
.PP
.B getpmsg()
reads a STREAMS message from the read queue of a stream head into the caller
supplied buffers.  The read messages will contain a control part or a data
part or both.  The data and control parts of the message are handled
separately by
.PP
The meaning and interpretation of the contents of the control part and data
part are specific to the STREAMS module to which the stream head is attached.
The semantics of these contents forms a necessary part of the STREAMS module
documentation.
.PP
.B getpmsg()
provides finer control over the messages retrieved than does
.BR getmsg (2).
.PP
.I fd
specifies a file descriptor referencing an open stream.
.PP
.I ctlptr
is used to hold the control part of the retrieved message and
.I datptr
is used to hold the data part of the retrieved message.
.IR ctlptr " and " dataptr
each contain
.B NULL
or a pointer to a
.B strbuf
structure, which contains the following members:
.sp
.nf
\fC\s-1\
struct strbuf {
    int maxlen;         /* maximum buffer length */
    int len;            /* length of data */
    char *buf;          /* ptr to buffer */
};
\s+1\fR
.fi
.TP 12
.I maxlen
specifies the maximum number of bytes that the buffer pointed to by
.I buf
can hold.
When 0 is specified,
.I maxlen
requests that only zero-length message be retrieved.
When -1 is specified,
.I maxlen
requests that the corresponding control or data part not be retrieved.
When set to a positive value,
.I maxlen
requests that at maximum only
.I maxlen
bytes of the control or data part be retrieved.
.TP 12
.I len
is a return value only.  See
.BR "\*(lqRETURN VALUES\*(rq" ,
below.
.TP 12
.I buf
points to the buffer into which the data or control information is to be placed.
.PP
If
.IR ctlptr " or " datptr
is
.BR NULL ,
or the corresponding
.I maxlen
field is set to -1, the control or data part of the message is not
processed and is left on the stream head read queue.
.PP
.I bandp
is the lowest numerical priority band of interest.  When
.I flagsp
is set to other than
.BR MSG_BAND ,
the integer pointed to by this argument must be coded zero (0); when
.I flagsp
is set to
.BR MSG_BAND ,
the integer pointed to by this argument indicates the lowest priority band of
interest.
.PP
.I flagsp
When
.I flagsp
points to an integer containing the flag
.BR MSG_ANY " or " MSG_HIPRI ,
.I bandp
should point to an integer containing the priority band 0.
When
.I flagsp
points to an integer containing the flag
.BR MSG_BAND ,
.I bandp
should point to an integer containing the lowest numerical priority band of
interest.
.I flagsp
should point to an integer that indicates the type of message
the user is able to receive and will return the type of message received.
.I flagsp
points to an integer that
can have one of the following values:
.TP 12
.BR MSG_HIPRI,
requests the retrieval of the first available high priority message
on the stream head read queue, and that low priority messages not be
processed.  For this option, the priority band indicated by
.I bandp
must be 0.
.TP 12
.BR MSG_ANY,
requests the retrieval of the first available message on the stream
head read queue, regardless of priority.  For this option, the priority band
indicated by
.I bandp
must be 0.
.TP 12
.BR MSG_BAND,
requests the retrieval of the first available message in a priority
band greater than or equal to the priority band specified in
.IR bandp .
.\"
.\"
.PP
If the stream head is set for non-blocking operation and no message of the
specified type and priority band is available to be read,
.B getpmsg()
will fail and set
.I errno
to
.BR EAGAIN " or " EWOULDBLOCK .
If the stream head is set for blocking operation,
.B getpmsg()
will block until a message of the specified type and priority band becomes
available on the stream head read queue, or the call is interrupted by a
signal.
If the call is interrupted by a signal in this fashion,
.B getpmsg()
will fail and set
.I errno
to
.BR EINTR .
.\"
.PP
If a hangup occurs on the stream from which messages are to be retrieved,
.B getpmsg()
continues to operate normally, as described above, until the
stream head read queue is empty.  Thereafter, it returns 0 in the
.I len
field of
.IR ctlptr " and " dataptr .
.\"
.\"
.SH "RETURN VALUES"
.PP
Upon failure,
.B getpmsg()
will
return -1 and set
.I errno
as described under
.BR "\*(lqERRORS\*(rq" ,
below.
.PP
Upon success,
.B getpmsg()
will return a non-negative value with any of the following flags set:
.TP 12
.B MORECTL
indicates that more control information belonging to the same message is
waiting to be retrieved from the stream head read queue with a subsequent
.B getpmsg()
operation.
.TP 12
.B MOREDATA
indicates that more data belonging to the same message is waiting to be
retrieved from the stream head read queue with a subsequent
.B getpmsg()
operation.
.B 0
indicates that the entire message was received successfully.
.PP
.B getpmsg()
with the appropriate arguments will retrieve a higher priority message before
the remainder of a partially retrieved lower priority message is retrieved.
.PP
Upon success,
.B getpmsg()
will also alter the
.I len
values in the
.B strbuf
structures pointed to by
.IR ctlbuf " and " datbuf
and the integer values pointed to by
.IR bandp " and " flagp ,
as follows:
The
.I maxlen
member is not altered on return.
If the
.IR ctlptr " or " datptr
a not
.BR NULL ,
and
.I maxlen
is not -1,
the existence and length of the retrieved message is returned in the
.I len
member.
If there is no control or data part retrieved, the
.I len
member of the corresponding
.B strbuf
structure is set to -1.
.PP
If the control or data part retrieved is of zero-length, the control or data
part is removed from the stream head read queue and the returned
.I len
is set to 0.
If
.I maxlen
is greater than or equal to 0 and the number of bytes in the control or
data part is greater than
.IR maxlen ,
only
.I maxlen
bytes are retrieved and the remaining bytes of the corresponding part are left
on the stream head read queue.  In the
.IR retval ,
the corresponding more flag
.BR "" ( MORECTL " or " MOREDATA )
is set.
The
.I buf
member is not altered on return.
.PP
When the integer that
.I flagp
points to returns the flag
.BR MSG_BAND ,
the integer
.I bandp
points to returns the priority band of the retrieved message.
.PP
For
.B getpmsg()
the integer pointed to by
.I flagsp
will contain one of the following values:
.TP 12
.BR MSG_HIPRI,
the retrieved message is a high priority message;
.TP 12
.BR MSG_BAND,
the retrieved message is not a high priority message--the value in the
integer pointed to by
.I bandp
indicates the priority band of the retrieved message.
.\"
.\"
.SH ERRORS
.PP
Upon failure,
.BR getpmsg()
will
return -1 and set
.I errno
as follows:
.TP 12
.RB [ EISDIR ]
.I fd
refers to a directory.
.TP 12
.RB [ EBADF ]
.I fd
is not a valid file descriptor or is not open for reading.
.TP 12
.RB [ EFAULT ]
A specified address is outside the user's accessible address space.
Specified addresses include,
.IR ctlptr ", " datptr ", " bandp ", " flagsp ,
and the
.I buf
fields of the
.B strbuf
structure pointed to by
.IR cttlptr " and " datptr .
.TP 12
.RB [ ENODEV ]
.I fd
refers to a device that does not support the
.B getpmsg()
system call.
.TP 12
.RB [ ENOSTR ]
.I fd
refers to a device that has no stream associated with it.
.TP 12
.RB [ EIO ]
I/O error, or,
.I fd
refers to a stream that is open but is in the process of closing.
.\".TP 12
.\".RB [ EPERM ]
.\".I fd
.\"points to a dynamic device and read permission on the device is denied.
.\"
.\"
.\"
.TP 12
.RB [ EINVAL ]
.I fd
is linked under a multiplexing driver; or,
.B getpmsg()
is not supported; or, the flags were invalid for
.B getpmsg()
(not
.BR MSG_HIPRI ", " MSG_ANY " or " MSG_BAND ).
.TP 12
.RB [ EAGAIN "], [" EWOULDBLOCK ]
The stream head is set for non-blocking operation and
.B getpmsg()
would block: that is, no message of the requested type and priority band is
available to be read.
.TP 12
.RB [ EINTR ]
A signal was caught, or an
.B M_SIG
message processed, before the call to
.B getpmsg()
could return any data.
.TP 12
.RB [ ENOSTR ]
.I fd
has no stream associated with the file descriptor.
.TP 12
.RB [ EBADMSG ]
The message at the head of the queue is of an invalid message type (valid
message types include:
.BR M_DATA ", " M_PROTO " and " M_PCPROTO );
or, the message at head of the of the queue is not of the requested type
as specified by
.IR flagsp ,
or priority band as specified by
.IR bandp .
.PP
Other errors may be returned by
.BR getpmsg() .
If the protocol module sends a
.B M_ERROR
message to the stream head, the error returned on all subsequent read
operations, including
.BR getpmsg() ,
is specified in the
.B M_ERROR
message by the protocol module.
Which errors are returned under which conditions form part of the
necessary documentation of the protocol module.
.\"
.\"
.SH CAVEATS
.PP
.B LiS
.[
LiS
.]
could handle concurrent readers better.  Linux sockets code blocks concurrent
readers while a partial reader is installed against a socket.  When partial
reads are performed for messages using
.BR getpmsg() .
.B LiS
.[
LiS
.]
should install the partial reader and block all other readers from completing
the partial read until the partial reader completes the read or the partial
reader closes the stream.  This would at least keep multiple threads of
execution from interfering with partial reads.
.\"
.\"
.SH NOTES
.PP
.B LiS
.[
LiS
.]
implements
.BR getmsg (2)
as a call to
.B getpmsg (2s)
with a
.B NULL
.I bandp
pointer.
.\"
.\"
.SH BUGS
.PP
.B LiS
.[
LiS
.]
incorrectly returns
.RB [ EINVAL ]
instead of
.RB [ EIO ]
when the internal head queue pointer points to a structure that is not a queue
structure, or is mangled.
.PP
.B LiS
.[
LiS
.]
incorreclty returns
.RB [ EINVAL ]
instead of
.RB [ EIO ]
when
.I q->q_first
yeilds a
.BR "" non- NULL
message pointer, but
.B getq()
later cannot retrieve a message.
.PP
.B LiS
.[
LiS
.]
does not return
.RB [ EIO ]
when
.B getpmsg()
is called on a closing stream.
.PP
.B LiS
.[
LiS
.]
incorrectly returns
.RB [ EAGAIN "] or [" EWOULDBLOCK ]
instead of
.RB [ EINTR ]
when an
.B M_SIG
message causes a signal that interrupts the call to
.B getpmsg()
before any data is read.
.PP
.B LiS
.[
LiS
.]
incorrectly returns
.RB [ ENODEV ]
instead of
.RB [ ENOSTR ]
when there is no stream associated with
.IR fd .
.PP
.B LiS
.[
LiS
.]
incorrectly returns
.RB [ EINVAL ]
instead of
.RB [ ENODEV ]
when
.I fd
refers to a device that does not support the
.B getpmsg()
system calls.
.\"
.\"
.SH "SEE ALSO"
.PP
.BR intro (2),
.BR poll (2),
.BR putmsg (2),
.BR read (2),
.BR write (2),
.BR streamio (2).
.\"
.\"
.SH "CONFORMING TO"
.PP
.BR SVID ", " SUSv2 ", " SUSv3 ", " POSIX .
.\"
.\"
.SH VERSIONS
.PP
This manpage was written for
.B LiS\c
.
.[
LiS
.]
.PP
.B getpmsg()
first appeared in SVR 3.
.[
svr3
.]
.\"
.\"
.XX
.[
magic
.]
.[
svr42
.]
.[
svr4
.]
.[
svr3
.]
.[
svid
.]
.[
xns
.]
.[
xbd5
.]
.[
$LIST$
.]
.TI


NEWS
OpenSS7 Master Package -- history of user visible changes.  2008-05-03
$Id: NEWS,v 0.9.2.53 2008-05-03 21:22:25 brian Exp $
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2000  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
OpenSS7 Master Package package.

 Major changes for release openss7-0.9.2.G
------------------------------------------

This is the sixth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Documented `lib32gcc1' problem on Ubuntu.

   - Noted problem running under XEN kernels.  XEN kernels are not yet
     supported.

   - License upgrade to AGPL v3.

   - Cross-compile fixes (strap out AC_FUNC_REALLOC macro when
     cross-compiling).

   - Ability to strap out major documentation build and installation
     primarily for embedded targets.

   - Improvements to common build process for embedded and
     cross-compile targets.

   - Support for flex 2.5.33  in maintainer mode.

   - Added the following sub-packages to the master build (and relase):

         - `strx25-0.9.2.1'


_This is a public stable production grade release of the package: it
deprecates previous releases.  Please upgrade to the current release
before reporting bugs._

As with other OpenSS7 releases, this release configures, compiles,
installs and builds RPMs and DEBs for a wide range of Linux 2.4 and 2.6
RPM- and DPKG-based distributions, and can be used on production
kernels without patching or recompiling the kernel.

This package is publicly released under the `GNU Affero General Public
License Version 3'.  The release is available as an `autoconf' tarball,
SRPM, DSC, and set of binary RPMs and DEBs.  See the downloads page
(http://www.openss7.org/download.html) for the `autoconf' tarballs,
SRPMs and DSCs.  For tarballs, SRPMs, DSCs and binary RPMs and DEBs,
see the openss7 package page (http://www.openss7.org/openss7_pkg.html).

See `http://www.openss7.org/codefiles/openss7-0.9.2.G/ChangeLog' and
`http://www.openss7.org/codefiles/openss7-0.9.2.G/NEWS' in the release
for more information.  Also, see the `openss7.pdf' manual in the
release (also in html `http://www.openss7.org/openss7_manual.html').

 Major changes for release openss7-0.9.2.F
------------------------------------------

This is the fifth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Support build on openSUSE 10.2.

   - Support build on Fedora 7 and 2.6.21 kernel.

   - Support build on CentOS 5.0 (RHEL5).

   - Support build on Ubuntu 7.04.

   - Updated to gettext 0.16.1.

   - Changes to support build on 2.6.20-1.2307.fc5 and
     2.6.20-1.2933.fc6 kernel.

   - Supports build on Fedora Core 6.

   - Support for recent distributions and tool chains.


 Major changes for release openss7-0.9.2.E
------------------------------------------

This is the fourth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

It is unlikely that the `OpenSS7 Master Package' will be released as
frequently as before.  Sub-packages will be released more often.  To
rebuild the master package with a new sub-package release, simply
replace the directory to which the sub-package belongs with the
unpacked sub-package release and then rebuild the master package.  This
release provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Addition of the `osr61' sub-package that contains Dialogic(R) Open
     System Release 6.1 version 239 GPL drivers.

   - A few minor corrections to the common build process.

   - Support for autoconf 2.61, automake 1.10 and gettext 0.16.

   - Support for Ubuntu 6.10 distribution and bug fixes for i386
     kernels.

   - The package now looks for subpackages with a version number as
     unpacked by separate tarball.


 Major changes for release openss7-0.9.2.D
------------------------------------------

This is the fourth public release of the OpenSS7 Master Package.  The
sub-packages have been reorganized for this release.  See `README' in
the release for a sub-package listing.  Aside from sub-package
reorganization, the major difference from previous release is that this
release no longer contains `LiS'.  Too many of the sub-packages will
not even build against `LiS' because of its Stream head deficiencies.


LiS/libc/getpmsg.c
/*****************************************************************************

 @(#) $RCSfile: getpmsg.c,v $ $Name:  $($Revision: 1.1.1.3.4.11 $) $Date: 2008-04-29 08:33:16 $

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified $Date: 2008-04-29 08:33:16 $ by $Author: brian $

 -----------------------------------------------------------------------------

 $Log: getpmsg.c,v $
 Revision 1.1.1.3.4.11  2008-04-29 08:33:16  brian
 - update headers for Affero release

 Revision 1.1.1.3.4.10  2007/08/14 10:47:13  brian
 - GPLv3 header update

 *****************************************************************************/

#ident "@(#) $RCSfile: getpmsg.c,v $ $Name:  $($Revision: 1.1.1.3.4.11 $) $Date: 2008-04-29 08:33:16 $"

static char const ident[] =
    "$RCSfile: getpmsg.c,v $ $Name:  $($Revision: 1.1.1.3.4.11 $) $Date: 2008-04-29 08:33:16 $";

/* This file can be processed with doxygen(1). */

#include "streams.h"

/** @addtogroup strcalls STREAMS System Calls
  * @{ */

/** @file
  * STREAMS System Call getpmsg() implementation file.  */

#ifdef BLD32OVER64
/**
 * @struct strbuf6
 * @date 2006-02-20
 * @author bidulock@openss7.org
 *
 * This is really, really, the wrong way to do this.  I cannot stress how wrong
 * this is.  This does not provide 32/64 bit compatibility at all.  It simply
 * attempts to turn a 32 bit call into a 64 bit call an passes it to the kernel.
 * A big problem is the pointer conversion (32->64), which is architecture
 * specific.  (Some need (void *)(long)(int32_t), some need (void
 * *)(ulong)(uint32_t), others might need to supply an offset.) For example of
 * how wrong this is: applications built on s390 using static libraries will not
 * run on s390x.  The proper way to do this is as it is done under Linux
 * Fast-STREAMS: leave the library alone and handle the conversion in the
 * kernel.  When performed by syscall, the syscall32 interface must perform the
 * conversion.  When performed by read/write call, the "magic length" must
 * indicate whether conversion is necessary.  When performed by ioctl, the 32/64
 * bit ioctl conversion functions must do the job.  I really have no desire to
 * "fix" LiS in this regard.  It is deprecated: don't use it.  Use Linux
 * Fast-STREAMS instead.
 */
#endif

#ifdef BLD32OVER64
typedef struct strbuf6 {
	int maxlen;
	int len;
	unsigned long long buf;
} strbuf6_t;

typedef struct getpmsg_args6 {
	int fd;
	unsigned long long ctl;
	unsigned long long dat;
	unsigned long long bandp;
	unsigned long long flagsp;

} getpmsg_args6_t;
#endif

int
__old_lis_getpmsg(int fd, struct strbuf *ctlptr, struct strbuf *datptr, int *bandp, int *flagsp)
{
#ifdef BLD32OVER64
	strbuf6_t ctl6;
	strbuf6_t *ptrc6 = 0;
	strbuf6_t dat6;
	strbuf6_t *ptrd6 = 0;
	getpmsg_args6_t args;
	int rc;

	if (ctlptr) {
		ctl6.maxlen = ctlptr->maxlen;
		ctl6.len = ctlptr->len;
		ctl6.buf = (unsigned long long) (unsigned long) ctlptr->buf;
		ptrc6 = &ctl6;
	}

	if (dataptr) {
		dat6.maxlen = dataptr->maxlen;
		dat6.len = dataptr->len;
		dat6.buf = (unsigned long long) (unsigned long) dataptr->buf;
		ptrd6 = &dat6;
	}

	memset((void *) &args, 0, sizeof(getpmsg_args6_t));
	args.fd = fd;
	args.ctl = (unsigned long long) (unsigned long) ptrc6;
	args.dat = (unsigned long long) (unsigned long) ptrd6;
	args.bandp = (unsigned long long) (unsigned long) bandp;
	args.flagsp = (unsigned long long) (unsigned long) flagsp;

	rc = read(fd, &args, LIS_GETMSG_PUTMSG_ULEN);

	if (ctlptr) {
		ctlptr->len = ctl6.len;
	}
	if (dataptr) {
		dataptr->len = dat6.len;
	}

	return (rc);
#else
	/**
	 * @date Wed Jun 22 20:56:59 MDT 2005
	 *
	 * This no longer works on FC4 2.6.11 kernel: validity checks are
	 * performed on the length before we get here.  We might as well patch
	 * this out for all kernels and use the ioctl method instead.  Emulating
	 * an system call in this fashion was foolish in the first place: the
	 * normal method for system call emulation under UNIX is with ioctl. -bb
	 */
	struct __lis_getpmsg_args {
		int fd;
		struct strbuf *ctlptr;
		struct strbuf *datptr;
		int *bandp;
		int *flagsp;
	};
	struct __lis_getpmsg_args args = {
		fd:fd,
		ctlptr:ctlptr,
		datptr:datptr,
		bandp:bandp,
		flagsp:flagsp,
	};

	return (read(fd, &args, LIS_GETMSG_PUTMSG_ULEN));
#endif
}

/** @brief get a message from a stream band.
  * @param fd a file descriptor representing the stream.
  * @param ctlptr a pointer to a strbuf structure returning the control part of the message.
  * @param datptr a pointer to a strbuf structure returning the data part of the message.
  * @param bandp a pointer to the band returned for the received message.
  * @param flagsp a pointer to the flags returned for the received message.
  */
int
__lis_getpmsg(int fd, struct strbuf *ctlptr, struct strbuf *datptr, int *bandp, int *flagsp)
{
#ifdef BLD32OVER64
	strbuf6_t ctl6;
	strbuf6_t *ptrc6 = 0;
	strbuf6_t dat6;
	strbuf6_t *ptrd6 = 0;
	getpmsg_args6_t args;
	int rc;

	if (ctlptr) {
		ctl6.maxlen = ctlptr->maxlen;
		ctl6.len = ctlptr->len;
		ctl6.buf = (unsigned long long) (unsigned long) ctlptr->buf;
		ptrc6 = &ctl6;
	}

	if (dataptr) {
		dat6.maxlen = dataptr->maxlen;
		dat6.len = dataptr->len;
		dat6.buf = (unsigned long long) (unsigned long) dataptr->buf;
		ptrd6 = &dat6;
	}

	memset((void *) &args, 0, sizeof(getpmsg_args6_t));
	args.fd = fd;
	args.ctl = (unsigned long long) (unsigned long) ptrc6;
	args.dat = (unsigned long long) (unsigned long) ptrd6;
	args.bandp = (unsigned long long) (unsigned long) bandp;
	args.flagsp = (unsigned long long) (unsigned long) flagsp;

	rc = read(fd, &args, LIS_GETMSG_PUTMSG_ULEN);

	if (ctlptr) {
		ctlptr->len = ctl6.len;
	}
	if (dataptr) {
		dataptr->len = dat6.len;
	}

	return (rc);
#else
	struct strpmsg args = { {-1, -1, NULL}, {-1, -1, NULL}, -1, -1 };
	int rc;

	if (ctlptr)
		args.ctlbuf = *ctlptr;
	if (datptr)
		args.databuf = *datptr;
	if (bandp)
		args.band = *bandp;
	if (flagsp)
		args.flags = *flagsp;
	if ((rc = ioctl(fd, I_GETPMSG, &args)) != -1) {
		if (ctlptr)
			ctlptr->len = args.ctlbuf.len;
		if (datptr)
			datptr->len = args.databuf.len;
		if (bandp)
			*bandp = args.band;
		if (flagsp)
			*flagsp = args.flags;
	}
	return (rc);
#endif
}

/** @brief get a message from a stream band.
  * @param fd a file descriptor representing the stream.
  * @param ctlptr a pointer to a strbuf structure returning the control part of the message.
  * @param datptr a pointer to a strbuf structure returning the data part of the message.
  * @param bandp a pointer to the band returned for the received message.
  * @param flagsp a pointer to the flags returned for the received message.
  * @version LIS_1.0 getpmsg()
  *
  * getpmsg() must contain a thread cancellation point (SUS/XOPEN/POSIX).  In the
  * Linux Threads approach, this function will return EINTR if interrupted by a
  * signal.  When the function returns EINTR, the Linux Threads user should check
  * for cancellation with pthread_testcancel().  Because this function consists
  * of a single system call, asynchronous thread cancellation protection is not
  * required.
  */
int
__lis_getpmsg_r(int fd, struct strbuf *ctlptr, struct strbuf *datptr, int *bandp, int *flagsp)
{
	int ret;

	pthread_testcancel();
	ret = __lis_getpmsg(fd, ctlptr, datptr, bandp, flagsp);
	if (ret == -1)
		pthread_testcancel();
	return (ret);
}

/** @brief get a message from a stream band.
  * @param fd a file descriptor representing the stream.
  * @param ctlptr a pointer to a strbuf structure returning the control part of the message.
  * @param datptr a pointer to a strbuf structure returning the data part of the message.
  * @param bandp a pointer to the band returned for the received message.
  * @param flagsp a pointer to the flags returned for the received message.
  * @version LIS_0.0 getpmsg()
  */
int
__old_lis_getpmsg_r(int fd, struct strbuf *ctlptr, struct strbuf *datptr, int *bandp, int *flagsp)
{
	int ret;

	pthread_testcancel();
	ret = __old_lis_getpmsg(fd, ctlptr, datptr, bandp, flagsp);
	if (ret == -1)
		pthread_testcancel();
	return (ret);
}

/** @fn int getpmsg(int fd, struct strbuf *ctlptr, struct strbuf *datptr, int *bandp, int *flagsp)
  * @brief get a message from a stream band.
  * @param fd a file descriptor representing the stream.
  * @param ctlptr a pointer to a strbuf structure returning the control part of the message.
  * @param datptr a pointer to a strbuf structure returning the data part of the message.
  * @param bandp a pointer to the band returned for the received message.
  * @param flagsp a pointer to the flags returned for the received message.
  * @version LIS_1.0 __lis_getpmsg_r()
  * @version LIS_0.0 __old_lis_getpsmg_r()
  */
__asm__(".symver __lis_getpmsg_r,getpmsg@@LIS_1.0");
__asm__(".symver __old_lis_getpmsg_r,getpmsg@LIS_0.0");

/** @} */

// vim: com=srO\:/**,mb\:*,ex\:*/,srO\:/*,mb\:*,ex\:*/,b\:TRANS
